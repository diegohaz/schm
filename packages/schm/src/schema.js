// @flow
import merge from 'lodash/merge'
import type { Schema, SchemaGroup } from './types'
import { isSchema } from './utils'
import parsers from './parsers'
import validators from './validators'
import parse from './parse'
import validate from './validate'
import parseParams from './parseParams'

export const defaultSchema = (params?: Object = {}): Schema => ({
  parsers,
  validators,
  parse(values) {
    return parse(values, this)
  },
  validate(values, paramPathPrefix) {
    return validate(values, this, paramPathPrefix)
  },
  merge(...schemas) {
    const merged = merge({}, this, ...schemas)
    return {
      ...merged,
      params: parseParams(merged.params),
    }
  },
  params: parseParams(params),
})

/**
 * A simple group of parameters. It's used internally when you pass literal
 * objects to [`schema`](#schema).
 * @example
 * const userSchema = schema(
 *   group({
 *     id: String,
 *     name: String,
 *   }),
 *   group({
 *     age: Number,
 *   })
 * )
 */
export const group = (params?: Object = {}): SchemaGroup => (
  (previous: Schema): Schema => previous.merge(defaultSchema(params))
)

/**
 * Creates a schema by composing groups of parameters.
 * @example
 * const userSchema = schema({
 *   id: String,
 *   name: String,
 * }, {
 *   age: Number
 * })
 *
 * // nested schema
 * const teamSchema = schema({
 *   users: [userSchema],
 * })
 */
const schema = (...groups: (Object | Schema | SchemaGroup)[]): Schema =>
  groups.reduce((finalSchema, currentGroup) => {
    if (typeof currentGroup === 'function') {
      return currentGroup(finalSchema)
    } else if (isSchema(currentGroup)) {
      return currentGroup
    }
    return group(currentGroup)(finalSchema)
  }, defaultSchema())

export default schema
