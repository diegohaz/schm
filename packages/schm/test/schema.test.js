import schema from '../src/schema'

test('parse', () => {
  const schm = schema({ foo: String })
  expect(schm.parse({ foo: 1 })).toEqual({ foo: '1' })
})

test('validate', async () => {
  const schm = schema({
    foo: {
      type: String,
      required: true,
    },
  })
  await expect(schm.validate()).rejects.toMatchSnapshot()
})

describe('composition', () => {
  it('composes schema group', () => {
    const foo = params => previous => previous.merge({
      params,
      parsers: {
        foo: (value, option) => `${value}${option}`,
      },
    })

    const bar = params => previous => previous.merge({
      params: Object.keys(params).reduce((finalParams, name) => ({
        ...finalParams,
        [name]: {
          bar: params[name],
        },
      }), {}),
      parsers: {
        bar: (value, option) => `${value}${option}`,
      },
    })

    expect(schema(
      {
        name: String,
      },
      foo({
        name: {
          foo: 'foo',
        },
      }),
      bar({
        name: 'bar',
      })
    ).parse({ name: 'test' })).toEqual({ name: 'testfoobar' })
  })

  it('composes schema', () => {
    const schema1 = schema({ age: Number })
    const schema2 = schema(schema1, { name: String })
    expect(schema2.parse({ name: 'Haz', age: '27' })).toEqual({
      name: 'Haz',
      age: 27,
    })
  })
})

describe('nested schema', () => {
  const parseMock = jest.fn()

  const studentSchema = schema({
    name: {
      type: String,
      required: true,
    },
    age: Number,
  }, previous => previous.merge({
    parse(values) {
      parseMock(values)
      return previous.parse(values)
    },
  }))

  const teacherSchema = schema({
    name: {
      type: String,
      required: true,
    },
  }, previous => previous.merge({
    parse(values) {
      parseMock(values)
      return previous.parse(values)
    },
  }))

  const classSchema = schema({
    name: {
      type: String,
      required: true,
    },
    teacher: {
      type: teacherSchema,
      required: true,
    },
    assistantTeacher: teacherSchema,
    students: {
      type: [studentSchema],
    },
    otherStudents: [studentSchema],
  })

  it('rejects validation', async () => {
    await expect(classSchema.validate()).rejects.toMatchSnapshot()
  })

  it('resolves validation', async () => {
    await expect(classSchema.validate({
      name: 'Computer Science',
      teacher: {
        name: 'Grace',
      },
    })).resolves.toMatchSnapshot()
  })

  it('parses', () => {
    expect(classSchema.parse()).toMatchSnapshot()
  })

  it('calls parse on nested schemas', () => {
    const teacher = { name: 'Grace' }
    const students = [{ name: 'foo' }, { name: 'bar' }]
    const otherStudents = [{ name: 'baz' }]
    parseMock.mockReset()
    expect(classSchema.parse({ teacher, students, otherStudents })).toMatchSnapshot()
    expect(parseMock).toHaveBeenCalledTimes(4)
    expect(parseMock).toHaveBeenCalledWith(teacher)
    expect(parseMock).toHaveBeenCalledWith(students[0])
    expect(parseMock).toHaveBeenCalledWith(students[1])
    expect(parseMock).toHaveBeenCalledWith(otherStudents[0])
  })
})
