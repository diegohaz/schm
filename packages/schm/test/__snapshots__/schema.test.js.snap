// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nested schema calls parse on nested schemas 1`] = `
Object {
  "assistantTeacher": undefined,
  "name": undefined,
  "otherStudents": Array [
    Object {
      "age": undefined,
      "name": "baz",
    },
  ],
  "students": Array [
    Object {
      "age": undefined,
      "name": "foo",
    },
    Object {
      "age": undefined,
      "name": "bar",
    },
  ],
  "teacher": Object {
    "name": "Grace",
  },
}
`;

exports[`nested schema parses 1`] = `
Object {
  "assistantTeacher": undefined,
  "name": undefined,
  "otherStudents": Array [],
  "students": Array [],
  "teacher": undefined,
}
`;

exports[`nested schema rejects validation 1`] = `
Array [
  Object {
    "message": "name is required",
    "param": "name",
    "required": true,
    "validator": "required",
    "value": undefined,
  },
  Object {
    "message": "teacher is required",
    "param": "teacher",
    "required": true,
    "validator": "required",
    "value": undefined,
  },
]
`;

exports[`nested schema resolves validation 1`] = `
Object {
  "assistantTeacher": undefined,
  "name": "Computer Science",
  "otherStudents": Array [],
  "students": Array [],
  "teacher": Object {
    "name": "Grace",
  },
}
`;

exports[`validate 1`] = `
Array [
  Object {
    "message": "foo is required",
    "param": "foo",
    "required": true,
    "validator": "required",
    "value": undefined,
  },
]
`;
